### ISSUE TITLE
simplification of cos(arccot(x) is wrong

### ISSUE THREAD

**Issue by hr87** [sympy/sympy#22070]
The simplification of `cos(arccot(x)` is according to [wikipedia](https://en.wikipedia.org/wiki/Inverse_trigonometric_functions#Relationships_between_trigonometric_functions_and_inverse_trigonometric_functions) 

```latex
cos(arccot(x) = x / sqrt{x^2 + 1}
```

But Sympy gives
```latex
cos(arccot(x) = 1 / sqrt{1/x^2 + 1}
```
which has a singularity, and problems with the sign for `x < 0`. The difference is shown in the plot below

![image](https://user-images.githubusercontent.com/12700929/132905225-9bed0dc4-b86b-48f2-b863-0f61237c6c7a.png)

**Comment by oscarbenjamin** [comment]
It should be straight-forward to change this:
https://github.com/sympy/sympy/blob/c4bdb03556b1bd9cf8b1e7465255d2a8a6c761ad/sympy/functions/elementary/trigonometric.py#L702-L704

**Comment by praneethratna** [comment]
Should'nt this also be changed then?
https://github.com/sympy/sympy/blob/3f8c8c2377cb8e0daaf8073e8d03ac7d87580813/sympy/functions/elementary/trigonometric.py#L368-L370
Since for sin also the plot is difference which is shown here,
![image](https://user-images.githubusercontent.com/63547155/132937896-eef0ea6f-213a-4a9c-9ce8-37df5db60912.png)

**Comment by oscargus** [comment]
Wolfram Alpha seems to support Sympy's version of both plot. Better change at Wikipedia...

**Comment by hr87** [comment]
Wolfram Alpha is also wrong.

We have on a unit circle
```
cot(x) = x/y       with     (0, pi) -> (-inf, inf)
```
(we have singularities at `0` and `pi` (`y=0`)), so the reverse function is then defined on `(-inf, inf) -> (0, pi)` without singularity. The cosine is defined on `(0, 2pi) -> (-1, 1)` without singularity. Chaining two functions, that have no singularities, and well overlapping intervals cannot give a singularity.

I assume Wolfram Alpha uses internally the `atan` function to express the `acot` as
```
acot(x) = atan(1/x)
```
However, `tan = y/x` is defined on `(-inf, inf) -> (-pi/2, pi/2), so `atan` is defined on `(-pi/2, pi/2) -> (-inf, inf)`, so this only works in the first quadrant `[0, pi/2]`, which is what we see in the plots. The actual expression should be
```
acot(x) = pi/2 - atan(x)     with    (-inf, inf) -> (0, pi)
```

**Comment by oscargus** [comment]
OK! I thought I'd make a survey for reference then:

### Maxima
![image](https://user-images.githubusercontent.com/8114497/132979099-fe4c7aac-e3de-46c7-86a8-693b9c9296b5.png)
### Matlab
![image](https://user-images.githubusercontent.com/8114497/132979155-f5e86414-56dc-4fff-8806-a552f8d8e3f1.png)
### Mathematica
![image](https://user-images.githubusercontent.com/8114497/132979192-ed255d86-5be3-48b4-9a18-d3e50455e308.png)
### Maple
![image](https://user-images.githubusercontent.com/8114497/132979330-964f21a5-0149-48e9-8d81-4ce6c2e2e84c.png)

I do not have any strong opinions, seems like a tie, but better to use these as arguments than Wikipedia.

**Comment by anutosh491** [comment]
Hi @oscargus ,I am an amateur/beginner in open source . I've only contributed to issues of graph/group theory modules of sage and networkx till now. I would like to work on the changes if there are strong enough reasons to make the change . Thank You!

**Comment by jksuom** [comment]
Since `cot` is a periodic function of period `pi`, its inverse `acot` is multivalued and `acot(x)` denotes only one of those values, the so called principal value. There is no universal way for determining the principal value, values in any interval of length `pi` can be chosen for real arguments.

In practice, there are two common choices for the range of principal values, either `[-pi/2, pi/2]`  or `[0, pi]`. SymPy and Mathematica take the former choice while Wikipedia uses the latter one. The values of `cos(acot(x))` agree when `acot(x)` is in the common part `[0, pi/2]` of the ranges but `cos` is positive in the interval `[-pi/2, 0]` and negative in `[pi/2, pi]`. Hence the rewritten expressions are different.

**Comment by oscarbenjamin** [comment]
In practice the definitions that SymPy takes for these things need to be consistent with mpmath and it's easy to see what choice mpmath makes:
```python
In [4]: import mpmath

In [5]: mpmath.cos(mpmath.acot(-1))
Out[5]: mpf('0.70710678118654757')
```

**Comment by anutosh491** [comment]
@oscarbenjamin 
Well that tells us clearly about the choice of the principle range, being used for `acot(x)` in sympy . But being new to contributing to sympy and also wanting to get more familiar with the code , I thought of checking the range using the `function_range` command.The range for `acot(x)` behaves differently than other trigonometric function in `function_range`. Not sure why is an empty set being returned .

![image](https://user-images.githubusercontent.com/87052487/132993978-18c4a12c-240e-401a-b554-0883b881bd34.png)

Is this a valid issue or am I missing something ? If valid , should I open a new issue for this !

**Comment by oscargus** [comment]
The empty set comes from `function_range` trying to find the point where the derivative of `acot` is zero.
`solveset(acot(x).diff(x), x)` returns `S.EmptySet`. This is in turn because of the discontinuity of `acot` at 0. `function_range` cannot handle discontinuous functions.

![image](https://user-images.githubusercontent.com/8114497/132998195-75225d91-a3c0-402d-8507-03f194999922.png)

**Comment by anutosh491** [comment]
Hello @oscargus ,I was a bit busy attending a kind off similar issue based on range/domain and inverse trigo functions . While looking for bugs for the other issue ,I realized that many a times switching these functions by their contemporary inverse trigo relations helps . eg `asin(x) = acsc(1/x)` and the other way round too . Generally speaking `acot(x) -atan(1/x) = 0` or `pi` but as in sympy we consider the choice of range of `acot(x)` to be `[-pi/2, pi/2]` ,`acot(x)` would be equal to` atan(1/x)` all the time .This also helps us in this case as shown below . 

I believe that this happens as `function_range` depends on `continuous domain` and both of these indeed ,depend on the `singularities` function .I strongly feel that the block/line of code responsible for this is  `for i in expression.rewrite([sec, csc, cot, tan], cos).atoms(Pow):`  in the singularities function ,cause the atoms(Pow) method returns different values for both contemporaries. I tried to get some info on passing Pow to atoms() but couldn't find much ,will try again. This helps us spot the singularity in the above case as `singularities(acot(x),x)` returns `EmptySet` but `singularities(atan(1/x),x)` returns `{0}` which is what we expect out of it . I'm not sure whether they should be returning different results or not.

![image](https://user-images.githubusercontent.com/87052487/133781249-cd6b41ef-d093-4cee-a3c2-03edca2e29bb.png)

But this doesn't always work in our favour as `singularities(asin(x),x)` should give us `EmptySet` as it is continuous throughout its domain but `singularities(acsc(1/x),x)` returns gives us `{0}` which might not expected in this case .Anyways, this error is more related to the second issue I have been trying to figure out (#21786 ) .
Coming to the improvement part . Being new to Sympy I would be glad if you could guide me through the changes or enhancements you would like to see in the` function_range` function.

**Comment by oscargus** [comment]
@anutosh491 unfortunately, I am not sure how to proceed here. It seems like there should be something similar to `singularities` called `discontinuities` (or something similar), but no idea if there is as I do not really know this part of the code that well.

Apart from that, I assume `function_range`, if no points where the derivative is zero is found, should try to find discontinuities and try to evaluate the limit from the left and the right and use those as part of determining the function range. Possibly even if there are points where the derivative is zero (in case a function has both, one can construct Piecewises with this for sure). But I do not really know enough about this part of SymPy to be able to tell what a feasible way to find discontinuities is.

**Comment by asmeurer** [comment]
There are several trig functions that have varying definitions. We should document better that there are multiple options and that we are picking a particular one, and reasoning behind it if there is any.

**Comment by anutosh491** [comment]
Yeah it's not that hard to find situations/use-cases where equal/equivalent trig functions may behave differently if you try . The comment I made above shows how inverse trig equivalents may behave differently when passed in `singularitities` , `continuous_domain` ,`function_range` .This could be inducing wrong results if we check for limits, differentials ,integrals including these (will have a check on that).
Your suggestion should surely be implemented . But not sure what all should we be tackling or how should we go about fixing this . If you have some suggestions in mind , could you elaborate ?

**Comment by gschintgen** [comment]
A few observations on this: A recent PR (#25948 ) improved `continuous_domain` and `singularities`. In particular:
```
In [9]: from sympy.calculus.util import continuous_domain, singularities

In [10]: continuous_domain(acot(x), x, Reals)
Out[10]: (-∞, 0) ∪ (0, ∞)

In [11]: singularities(acot(x), x, Reals)
Out[11]: ∅
```
Plotting should now work better for `acot` (I haven't checked). The definition chosen by SymPy is succinctly explained in its documentation.

As for detecting discontinuities by limit computation on both sides, that's unfortunately not possible yet. See #24127, #23836. I'd very much like to see (hypothetical) functions like `discontinuities()` and `is_continuous_at()`  in `sympy.calculus` but for the time being that probably isn't really possible or at least it would have to be restricted to those discontinuous functions that don't involve a `Piecewise`.
